bson/src/main/org/bson/codecs/pojo/PojoCodecImpl.java

```diff
@@ -141,22 +142,53 @@ public String toString() {
         return classModel;
     }
 
+
+    private LevelCountingBsonWriter wrapBsonWriter(final BsonWriter writer) {
+        if (writer instanceof LevelCountingBsonWriter) {
+            return (LevelCountingBsonWriter) writer;
+        }
+        return new LevelCountingBsonWriter(writer);
+    }
+
+    private <S> void encodeIdProperty(final LevelCountingBsonWriter writer, final T instance, final EncoderContext encoderContext,
+                                      final IdPropertyModelHolder<S> propertyModelHolder) {
+        if (propertyModelHolder.getPropertyModel() == null || propertyModelHolder.getIdGenerator() == null) {
```

Comment:
I found this conditional confusing, since if the property model is null, encodeProperty is a no-op.  How about?

        if (propertyModelHolder.getPropertyModel() != null) {
            if (propertyModelHolder.getIdGenerator() == null) {
                encodeProperty(writer, instance, encoderContext, propertyModelHolder.getPropertyModel());
            } else {
                 S id = ...
           }
    }

Reply:
Done